libminifi/src/processors/ContentHash.cpp

```diff
@@ -0,0 +1,100 @@
+/**
+ * @file ContentHash.cpp
+ * ContentHash class implementation
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifdef OPENSSL_SUPPORT
+
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <string>
+#include "processors/ContentHash.h"
+#include "core/ProcessContext.h"
+#include "core/ProcessSession.h"
+#include "core/FlowFile.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace processors {
+
+core::Property ContentHash::HashAttribute("Hash Attribute", "Attribute to store checksum to", "Checksum");
+core::Property ContentHash::HashAlgorithm("Hash Algorithm", "Name of the algorithm used to generate checksum", "MD5");
+core::Relationship ContentHash::Success("success", "success operational on the flow record");
+
+void ContentHash::initialize() {
+  //! Set the supported properties
+  std::set<core::Property> properties;
+  properties.insert(HashAttribute);
+  properties.insert(HashAlgorithm);
+  setSupportedProperties(properties);
+  //! Set the supported relationships
+  std::set<core::Relationship> relationships;
+  relationships.insert(Success);
+  setSupportedRelationships(relationships);
+}
+
+void ContentHash::onTrigger(core::ProcessContext *context, core::ProcessSession *session) {
+  std::shared_ptr<core::FlowFile> flowFile = session->get();
+
+  if (!flowFile) {
+    return;
+  }
+
+  ReadCallback cb(flowFile, context);
+  session->read(flowFile, &cb);
+  session->transfer(flowFile, Success);
+}
+
+int64_t ContentHash::ReadCallback::process(std::shared_ptr<io::BaseStream> stream) {
+  std::string attrKey, algoName;
+  ctx_->getProperty(HashAttribute.getName(), attrKey);
+  ctx_->getProperty(HashAlgorithm.getName(), algoName);
+  std::transform(algoName.begin(), algoName.end(), algoName.begin(), ::toupper);
+
+  // Erase '-' to make sha-256 and sha-2 work, too
+  algoName.erase(std::remove(algoName.begin(), algoName.end(), '-'), algoName.end());
+
+  // This throws in case algo is not found, but that's fine
```

Comment:
This is definitely a interesting point. 
I would consider failure relationship being user for flowfile specific issues mostly. For eg. the processor expects some data to be present in attributes/content of the flowfile, and this criteria isn't met. 
Although this case is about the processor being misconfigured, which in my view is a bit different. In an ideal world this shoudn't even happen as setting the property should fail in case the provided value is not one of the allowed ones.

Reply:
Failure is a generic term and is generally up to the independent processor to define -- your definition is not true across all processors in NiFi, so I'm fine leaving that up to the author to decide in this case; however, if the empty content case accounted for within the processor? Some have perceived empty content hashing as a potential failure case. Can make that a follow on task, though, but we should probably reach parity with NiFI eventually